cmd/download_headers.cpp

```diff
@@ -120,6 +123,7 @@ int main(int argc, char* argv[]) {
 
     // wait threads termination
     if (message_receiving.joinable()) message_receiving.join();
+    if (stats_receiving.joinable()) message_receiving.join();
     if (block_request_processing.joinable()) block_request_processing.join();
     if (header_processing.joinable()) header_processing.join();
```

Comment:
Thread joinable() is in most cases un-needed as it only checks thread_id != 0.
I'd move the join to threads inside the above try catch block and join without condition so we can intercept exceptions thrown by the implementation which is :
```
   void join() {
        if (!joinable()) {
            _Throw_Cpp_error(_INVALID_ARGUMENT);
        }

        if (_Thr._Id == _Thrd_id()) {
            _Throw_Cpp_error(_RESOURCE_DEADLOCK_WOULD_OCCUR);
        }

        if (_Thrd_join(_Thr, nullptr) != _Thrd_success) {
            _Throw_Cpp_error(_NO_SUCH_PROCESS);
        }

        _Thr = {};
    }
```
```

Reply:
If an exception is raised before all the four threads are constructed than some of them will be NON joinable. 
Anyway in this condition also the stop() methods would not be called so the program would hang.  We need a stop() & join() from the destructor of BlockProvider, HeaderDownloader & SentryClient.

As said before here, right now the downloader is like an engine that still needs to be installed on the platform. When we will connect the downloader with the stage-sync loop we will address those issues. I think it is a waste of time to correctly address here those corner cases and then re-do the work when we will connect the downloader with the  stage-sync loop that has, IMO, different needs.  What do yuo think?

Anyway I am deleting the ` if (...joinable())`